### DISCUSSION TITLE
Compile to add Subscribe? and other questions

### DISCUSSION THREAD

**Discussion by TinajaLabs** [arendst/Tasmota#10300] [Q&A] [answered]
I see a note in the docs that says I have to compile my own version of Tasmota to include the MQTT Subscribe command.

I see there's a header file I must adjust and I'll stumble through that.

I'm brand new to Tasmota.  My main device is the Wemos D1-mini and I've used the arduino IDE and various libraries (like pub-sub) to make my things.

I'm trying to duplicate this basic setup... using 2 Wemos D1-minis and a node-red/mosquito broker:
1. a motion sensor publishes state
2. a strand of LEDs subscribes to that state

Details:
1. First device is a pir sensor with the hi/lo pin hooked to the D6/GPIO12 pin - this publishes to the mqtt server with topic `home/bedroom/pir/state/wemosd1-1`.
2. Second device is a short strand of neopixel leds with the control line connected to the D2/GPIO4 pin - it subscribes to my mqtt server with topic `home/bedroom/pir/state/wemosd1-1`.

I've got the first device, the motion sensor, responding to movement.  And I have a rule to publish an ON value to the topic.  (still trying to understand how to Subscribe on the second device).

```
Rule1
  ON switch1#state==1 DO publish home/bedroom/pir/state/wemosd1-1 {"state":"%value%"} ENDON
  ON switch1#state==0 DO publish home/bedroom/pir/state/wemosd1-1 {"state":"%value%"} ENDON
```

When I wiggle my hand in from of the PIR sensor, I see this in the console:
```
03:09:50 RUL: SWITCH1#STATE==1 performs "publish home/bedroom/pir/state/wemosd1-1 {"state":"1"}"
03:09:50 MQT: home/bedroom/pir/state/wemosd1-1 = {"state":"1"}
```
I never see the 0 value (low/gnd state) five seconds after the last time the GPIO12 pin went high.

Questions:
1. On the PIR sensor I can see the 1/ON value get to the mqtt server but I never see the 0/off value.  How do I deliver MQTT msgs when it goes high AND when it goes low?
2. Looks like I must perform a special compile to include the Subscribe command.  This seems to indicate the pub/sub is not much used.  Is there a better way, or am I missing something?  
3. And if I do what's a typical rule to turn on a strand of neopixel LEDs based on subscription?

Thanks for any advice,
Chris.

**Comment by barbudor** [comment]
Hello @TinajaLabs
Edit: sorry you are on Discussions, not Issues. Please accept my apologies. 

For general support please join us on Discord (see link on main Readme) 

Pub/sub is at the core of Tasmota. The additional feature you are referring is 'generic subscribing to any topic from a Rule or a Script.
Which is not mandatory for your use case as I understand it. 
Tasmota already include various means for devices to communicate and synchronize through MQTT. 

Let's discuss this on Discord

**Comment by barbudor** [comment, marked as answer]
Hi @TinajaLabs 

As said above, there are multiple ways to reach your goal

1) Publishing directly from PIR to Leds on standard topic
Your PIR module could publish directly to the LED module on a standard topic like `Publish cmnd/topicled/POWER on`
You need Rules only on the PIR moduleOr eve use the command ` SwitchTopic` to have the PIR module to natively publish to a given custom topic

2) Publishing directly from PIR to leds on an EVENT
Your PIR module could publish directly to the LED module on an evet topic like `Publish cmnd/topicled/EVENT pirevent`
You also need Rules on the led module such as ` on Event#pirevent DO whateveryouwant endon`

3) Publishing and subscribing to a custom generic event
This feature is not cmpiled in by default in standard build
You can either use the additional binary 'tasmota-fullrules.bin(.gz)` or compile your own
To add support for generic topic subscription for Rumes or Script, you need to enable SUPPORT_MQTT_EVENT in the `user_config_override.h` header

4) Take advantage of NodeRed
At last you can also use NodeRed to translate a topic from PIR module into commands for the LED module

I hope this gives you some directions
More may be available too

Sorry again for my 1st message, I'm not yet used to Discussions.
Let's meet on Discord for mor eintercative discussion

**Comment by TinajaLabs** [comment]
Thanks, Barbudor,

I'll transition to Discord.  This, Discussions, seemed like a better place to lay out my setup and conditions, but maybe Discord will work as well.  I am busy this morning but will get back with more questions later.

**Comment by barbudor** [comment]
Hi @TinajaLabs 
Any more help needed ?

**Comment by blakadder** [comment]
Use -fullrules binary on https://github.com/tasmota/binaries/. It has all the rule defines enabled, including MQTT subscribe.
